pkg/cmd/cli/secrets/source.go

```diff
@@ -0,0 +1,230 @@
+		Long:    createSourceSecretLong,
+		Example: fmt.Sprintf(createSourceSecretExample, fullName, newSecretFullName, ocEditFullName),
+		Run: func(c *cobra.Command, args []string) {
+			if err := o.Complete(f, args); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := o.Validate(); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+
+			if len(kcmdutil.GetFlagString(c, "output")) != 0 {
+				secret, err := o.MakeSourceSecret()
+				kcmdutil.CheckErr(err)
+
+				kcmdutil.CheckErr(f.PrintObject(c, secret, out))
+				return
+			}
+
+			if err := o.CreateSourceSecret(); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	cmd.Flags().StringVar(&o.Username, "username", "", "username for Git authentication")
+	cmd.Flags().StringVar(&o.Password, "password", "", "password or token for Git authentication")
+	cmd.Flags().StringVar(&o.PrivateKeyPath, "ssh-privatekey", "", "path to the ssh private key")
+	cmd.Flags().StringVar(&o.CertificatePath, "ca-cert", "", "path to the certificate file")
+	cmd.Flags().StringVar(&o.GitConfigPath, "gitconfig", "", "path to the .gitconfig file")
+	cmd.Flags().BoolVarP(&o.PromptForPassword, "prompt", "", false, "prompt for password or token")
+	kcmdutil.AddPrinterFlags(cmd)
+
+	return cmd
+}
+
+func (o *CreateSourceSecretOptions) CreateSourceSecret() error {
+	secret, err := o.MakeSourceSecret()
+	if err != nil {
+		return err
+	}
+
+	if _, err := o.SecretsInterface.Create(secret); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(o.GetOut(), "secret/%s\n", secret.Name)
+	return nil
+}
+
+func (o *CreateSourceSecretOptions) MakeSourceSecret() (*api.Secret, error) {
+	secret := &api.Secret{}
+	secret.Namespace = o.SecretNamespace
+	secret.Name = o.SecretName
+	secret.Type = buildapi.SecretTypeSource
+	secret.Data = map[string][]byte{}
+
+	if len(o.Username) != 0 {
+		secret.Data[buildapi.SourceUsername] = []byte(o.Username)
+	}
+
+	if len(o.Password) != 0 {
+		secret.Data[buildapi.SourcePassword] = []byte(o.Password)
+	}
+
+	if len(o.CertificatePath) != 0 {
+		caContent, err := ioutil.ReadFile(o.CertificatePath)
+		if err != nil {
+			return nil, err
+		}
+		secret.Data[buildapi.SourceCertificate] = caContent
+	}
+
+	if len(o.GitConfigPath) != 0 {
+		gitConfig, err := ioutil.ReadFile(o.GitConfigPath)
+		if err != nil {
+			return nil, err
+		}
+		secret.Data[buildapi.SourceGitconfig] = gitConfig
+	}
+
+	if len(o.PrivateKeyPath) != 0 {
+		privateKeyContent, err := ioutil.ReadFile(o.PrivateKeyPath)
+		if err != nil {
+			return nil, err
+		}
+		secret.Data[buildapi.SourcePrivateKey] = privateKeyContent
+	}
+
+	return secret, nil
+}
+
+func (o *CreateSourceSecretOptions) Complete(f *kcmdutil.Factory, args []string) error {
+	if len(args) != 1 {
+		return errors.New("must have exactly one argument: secret name")
+	}
+	o.SecretName = args[0]
+
+	if len(o.Username) != 0 || len(o.Password) != 0 || len(o.CertificatePath) != 0 {
```

Comment:
Actually `Complete` quite often needs to overlap validation which makes it difficult to separate between the two. I think this is ok to stay here since o.Password is involved enough below